### PR TITLE
Handle lost state due to overflow and unknown events

### PR DESCRIPTION
**Pull request by lptr**
This adds handling for file system watcher overflows.

It bumps `native-platform` to `0.20-snapshot-20191211211216+0000`. That incorporates changes from this PR: https://github.com/adammurdoch/native-platform/pull/52

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/internal/vfs/impl/WatcherEvent.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        }
+    }
+
+    class ChangeEvent implements WatcherEvent {
+        private final FileWatcherRegistry.Type type;
+        private final Path path;
+
+        public ChangeEvent(FileWatcherRegistry.Type type, String path) {
+            this.type = type;
+            this.path = Paths.get(path);
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/internal/vfs/impl/WatcherEvent.java]: Why are we converting the path `String` into a `Path` only to convert it back again when talking to the VFS?

Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/impl/WatcherEvent.java]: Maybe that's future proofing too much, but I'd like to use `Path` as the API. If it proves to be too much of a burden, we can change it to use plain `String`s. I'd keep it this way for now.

Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/impl/WatcherEvent.java]: That is fine with me. Currently, this is not performance critical.

File: subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultWatchingVirtualFileSystem.java [hunk 2]
```diff
@@ -108,32 +107,32 @@ public void stopWatching() {
...
 
         AtomicInteger count = new AtomicInteger();
-        AtomicBoolean overflow = new AtomicBoolean();
+        AtomicBoolean unknownEventEncountered = new AtomicBoolean();
         try {
             watchRegistry.stopWatching(new FileWatcherRegistry.ChangeHandler() {
                 @Override
                 public void handleChange(FileWatcherRegistry.Type type, Path path) {
                     count.incrementAndGet();
+                    LOGGER.debug("Handling VFS change {} {}", type, path);
```
Comment (wolfs) [inline subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultWatchingVirtualFileSystem.java]: Can we move this to info logging while the feature is in the early phase? Or directly log it to the daemon log. WDYT?

Running with debug output on a bigger project is very painful.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/impl/WatcherEvent.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            case UNKNOWN:
+                return StateLostEvent.INSTANCE;
+            default:
+                throw new AssertionError();
+        }
+    }
+
+    static void dispatch(Iterable<WatcherEvent> events, FileWatcherRegistry.ChangeHandler handler) throws IOException {
+        try {
+            events.forEach(watcherEvent -> watcherEvent.dispatch(handler));
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/internal/vfs/impl/WatcherEvent.java]: I just got a `java.util.ConcurrentModificationException` at this point. This probably means we need to close the watcher first, so we don't get any more changes to the list while we are iterating it. Another option is to use a concurrency safe data structure, like a `CopyOnWriteArrayList`.

**Pull request merged**